// _@ali_ding-filetypes@0.1.5@@ali/ding-filetypes/index.js
var TYPE_IDS = {
}

var ID_CATE_TYPES = {
    0: 'IMAGE_JPG',
    1:'IMAGE_GIF',
    2: 'IMAGE_PNG',
    3:'IMAGE_BMP',
    4:'AUDIO_AMR',
    5:'AUDIO_MP3',
    6:'VIDEO_MP4',
    7:'AUDIO_WAV',
    8:'NORMAL_FILE',
    9:'OFFICE_DOC',
    10:'OFFICE_DOCX',
    11:'OFFICE_XLS',
    12:'OFFICE_XLSX',
    13:'OFFICE_PPT',
    14:'OFFICE_PPTX',
    15:'NORMAL_ZIP',
    16:'NORMAL_PDF',
    17:'NORMAL_RAR',
    18:'NORMAL_PSD',
    19:'NORMAL_AI',
    20:'NORMAL_TXT',
    21:'VIDEO_AVI',
    22:'VIDEO_RMVB',
    23:'VIDEO_RM',
    24:'VIDEO_MPG',
    25:'VIDEO_WMV',
    26:'VIDEO_MKV',
    27:'VIDEO_VOB',
    28:'NORMAL_TFSPRIVATE',
    29:'IMAGE_WEBP',
    30:'NORMAL_PAGES',
    31:'NORMAL_NUMBERS',
    32:'NORMAL_KEY',
    33:'NORMAL_SKETCH'
};

Object.keys(ID_CATE_TYPES).forEach(function(typeId){
    var cateType = ID_CATE_TYPES[typeId];
    var type = cateType.split('_')[1].toLowerCase();
    TYPE_IDS[type] = typeId;
});
TYPE_IDS['jpeg'] = 0;

var VALID_TYPES = ["jpg", "jpeg", "bmp", "png", 'gif', "doc", "docx", "xlsx", "xls", "pdf", "ppt", "pptx", "zip", "rar", 'txt', 'mp4', 'psd', 'ai', 'avi'];

var typeFuncs = {
    getTypeByName : function(fname){
        return fname.split('.').pop().toLowerCase();
    },
    getIdByType : function(type){
        if(TYPE_IDS[type] === undefined){
            var type = 'file';
        }
        return TYPE_IDS[type];
    },
    getCateTypeById : function(typeId){
        if(ID_CATE_TYPES[typeId] === undefined){
            var typeId = '8';
        }
        return ID_CATE_TYPES[typeId];
    },
    getTypeById : function(typeId){
        if(ID_CATE_TYPES[typeId] === undefined){
            var typeId = '8';
        }
        return ID_CATE_TYPES[typeId].split('_')[1].toLowerCase();
    },
    getCateTypeByType : function(type){
        return typeFuncs.getCateTypeById(typeFuncs.getIdByType(type));
    },
    isValid : function(type){
        var index = VALID_TYPES.indexOf(type);
        return index !== -1;
    },
    isImg:function(type){
        return typeFuncs.getCateTypeByType(type).split('_')[0] === 'IMAGE';
    },
    getFileIconByType:function(type){
        var cateType = typeFuncs.getCateTypeByType(type);
        switch(cateType){
            case 'OFFICE_DOC':
            case 'OFFICE_DOCX':
                return 'word-icon';
            case 'OFFICE_PPT':
            case 'OFFICE_PPTX':
                return 'ppt-icon';
            case 'NORMAL_PDF':
                return 'pdf-icon';
            case 'OFFICE_XLS':
            case 'OFFICE_XLSX':
                return 'excl-icon';
            case 'NORMAL_ZIP':
                return 'zip-icon';
            case 'NORMAL_RAR':
                return 'rar-icon';
            case 'NORMAL_PSD':
                return 'psd-icon';
            case 'NORMAL_AI':
                return 'ai-icon';
            case 'NORMAL_TXT':
                return 'txt-icon';
            case 'NORMAL_KEY':
                return 'key-icon';
            case 'NORMAL_PAGES':
                return 'pages-icon';
            case 'NORMAL_SKETCH':
                return 'sketch-icon';
            case 'NORMAL_NUMBERS':
                return 'numbers-icon'
            case 'VIDEO_AVI':
            case 'VIDEO_RMVB':
            case 'VIDEO_RM':
            case 'VIDEO_MPG':
            case 'VIDEO_WMV':
            case 'VIDEO_MKV':
            case 'VIDEO_VOB':
            case 'VIDEO_MP4':
                return 'video-icon';
            case 'IMAGE_JPG':
            case 'IMAGE_GIF':
            case 'IMAGE_PNG':
            case 'IMAGE_BMP':
                return 'pic-icon';
            default:
                return 'default-icon';
        }
    }
}

module.exports = typeFuncs;
